File: unity/Assets/Script/Comp/GameCanvas.cs
/**
 * Copyright 2022. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using static CloudList<FrameSync.Cloud>;
using static PlayerList<FrameSync.Player>;

public class GameCanvas : MonoBehaviour {

    // 初始化玩家对象池
    public List<GameObject> playersPool = new List<GameObject>();
    // 初始化云朵对象池
    Dictionary<int, GameObject> cloudDictionnary = new Dictionary<int, GameObject>();

    // 初始化子弹对象池
    public List<GameObject> bulletsPool = new List<GameObject>();

    // 初始化圆圈对象池
    public List<GameObject> circlesPool = new List<GameObject>();
    // 玩家预制件
    public GameObject playerPrefab;

    // 云朵预制件
    public GameObject cloudPrefab;

    // 子弹预制件
    public GameObject bullectPrefab;
    
    // 圆圈预制件
    public GameObject circlePrefab;
    
    public void SetCircle(bool circleDisplay)
    {
        if (circlePrefab == null) {
            return;
        }
        if (circleDisplay)
        {
            if (circlesPool.Count <= 0)
            {
                GameObject circleGameObject = Instantiate(circlePrefab);
                circlesPool.Add(circleGameObject);
                Circle circle =circlesPool[0].GetComponent<Circle>();
                circle.initCircle();
            }
            
        }
        else
        {
            if (circlesPool.Count <= 0) {
                return;
            }
            foreach (GameObject circle in circlesPool)
            {
                Destroy(circle);
            }
            circlesPool.Clear();
        }
    }
    
    public void SetPlayers(List<PlayerData<FrameSync.Player>> players)
    {
        if (playerPrefab == null) {
            return;
        }
        if (players == null || players.Count <= 0) {
            return;
        }
        if (playersPool.Count != players.Count) {
            // 清空预制件
            if (playersPool.Count != 0)
            {
                foreach (GameObject player in playersPool)
                {
                    Destroy(player);
                }
                playersPool.Clear();
            }
            // 重新初始化预制件
            for (int i = 0; i < players.Count; i++) {
                GameObject playerGameObject = Instantiate(playerPrefab);
                playersPool.Add(playerGameObject);
            }
        }
        // 绘制玩家
        for (int index = 0; index < players.Count; index++) {
            PlayerData<FrameSync.Player> player = players[index];
            Players playerView = playersPool[index].GetComponent<Players>();
            playerView.InitPlayer(player.id, player.rotation, player.x, player.y, player.playerTeamId);
        }

    }

    public void SetClouds(List<CloudData<FrameSync.Cloud>> clouds, float dt)
    {
        if (cloudPrefab == null)
        {
            return;
        }
        if (clouds == null || clouds.Count <= 0)
        {
            return;
        }
        // 绘制云朵
        for (int index = 0; index < clouds.Count; index++)
        {
            // 初始化预制件
            CloudData<FrameSync.Cloud> cloud = clouds[index];
            if (!cloudDictionnary.ContainsKey(index)) {
                GameObject cloudGameObject = Instantiate(cloudPrefab);
                cloudDictionnary.Add(index, cloudGameObject);
            }
            float offset = cloud.x + cloud.offset;
            // 根据位置绘制云朵
            if (offset > FrameSync._maxX)
            {
                // 如果超出X轴就销毁预制件
                GameObject cloudGameObject = cloudDictionnary[index];
                Destroy(cloudGameObject);
                cloudDictionnary.Remove(index);
            }
            else
            {
                // 如果没有超过最大X轴，拿到预制件进行绘制云朵
                GameObject cloudGameObject = cloudDictionnary[index];
                Clouds cloudView = cloudGameObject.GetComponent<Clouds>();
                cloud.offset += cloud.speed * dt;
                cloudView.InitCloud(offset, cloud.y);
            }
        }
    }

    public void setBullets(List<BulletList<FrameSync.Bullet>.BulletData<FrameSync.Bullet>> bullets)
    {
        if (bullectPrefab == null)
        {
            return;
        }
        if (bullets == null || bullets.Count < 0)
        {
            return;
        }
        if (bulletsPool.Count != bullets.Count || bullets.Count == 0) {
            // 清空预制件
            if (bulletsPool.Count != 0)
            {
                foreach (GameObject bullet in bulletsPool)
                {
                    Destroy(bullet);
                }
                bulletsPool.Clear();
            }
            // 重新初始化预制件
            for (int i = 0; i < bullets.Count; i++) {
                GameObject bulletGameObject = Instantiate(bullectPrefab);
                bulletsPool.Add(bulletGameObject);
            }
        }

        // 绘制子弹
        for (int index = 0; index < bullets.Count; index++) {
            BulletList<FrameSync.Bullet>.BulletData<FrameSync.Bullet> bullet = bullets[index];
            Bullet bulletView = bulletsPool[index].GetComponent<Bullet>();
            bulletView.initBullet(bullet.x, bullet.y, bullet.playerId, bullet.bulletId);
        }
    }

}

File: cocos/assets/Script/comp/GameCanvas.ts
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2012-2017 DragonBones team and other contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 *  2021.12.15-Changed method initPlayersPool
 *  2021.12.15-Changed method setPlayers
 *  2023.06.28-Changed method getFromPlayersPool
 *  2023.06.28-Deleted method initBullectPool
 *  2023.06.28-Deleted method getFromBullectPool
 *  2023.06.28-Changed method start
 *  2023.06.28-Changed method setClouds
 *  2023.06.28-Changed method setPlayers
 *  2023.06.28-Changed method setBullet
 *  2023.06.28-Add method destroyBullet
 *             Copyright(C)2023. Huawei Technologies Co., Ltd. All rights reserved
 */

// Learn TypeScript:
//  - [Chinese] https://docs.cocos.com/creator/manual/zh/scripting/typescript.html
//  - [English] http://www.cocos2d-x.org/docs/creator/manual/en/scripting/typescript.html
// Learn Attribute:
//  - [Chinese] https://docs.cocos.com/creator/manual/zh/scripting/reference/attributes.html
//  - [English] http://www.cocos2d-x.org/docs/creator/manual/en/scripting/reference/attributes.html
// Learn life-cycle callbacks:
//  - [Chinese] https://docs.cocos.com/creator/manual/zh/scripting/life-cycle-callbacks.html
//  - [English] http://www.cocos2d-x.org/docs/creator/manual/en/scripting/life-cycle-callbacks.html
import {PlayerData} from "../function/PlayerList";
import Player from "./Player";
import Cloud from "./Cloud";
import {CloudData} from "../function/CloudList";
import {BulletData} from "../function/BulletList";
import Bullet from "./Bullet";
import {
    destroyedBulletSet,
    frameSyncPlayerInitList
} from "../function/FrameSync";
import global from "../../global";

const {ccclass, property} = cc._decorator;
let playersPool: cc.NodePool = null;
let cloudsPool: cc.NodePool = null;
// 初始化对象池
function initPlayersPool(playerPrefab: cc.Prefab) {
    if (playersPool) {
        return;
    }
    playersPool = new cc.NodePool();
    let player = cc.instantiate(playerPrefab);
    playersPool.put(player);
}

function getFromPlayersPool(playerPrefab: cc.Prefab) {
    /*let player;
    if (playersPool.size() > 0) {
        player = playersPool.get();
    } else {
        player = cc.instantiate(playerPrefab);
    }

    return player;*/
    let player = cc.instantiate(playerPrefab);
    return player;
}

function getFromCloudsPool(cloudPrefab: cc.Prefab) {
    let cloud;
    if (cloudsPool != null && cloudsPool.size() > 0) {
        cloud = cloudsPool.get();
    } else {
        cloud = cc.instantiate(cloudPrefab);
    }
    return cloud;
}

function removeToPlayerPool(player) {
    playersPool.put(player);
}

@ccclass
export default class GameCanvas extends cc.Component {

    @property(cc.Prefab)
    playerPrefab: cc.Prefab = null;

    @property(cc.Prefab)
    cloudPrefab: cc.Prefab = null;

    @property(cc.Prefab)
    circlePrefab: cc.Prefab = null;

    @property(cc.Prefab)
    bulletPrefab: cc.Prefab = null;

    public players: cc.Node[] = [];
    public clouds: Cloud[] = [];
    public tileSize = 40;
    public cloudSize = 36;
    public maxX = 19; // x轴最大值

    start() {
        initPlayersPool(this.playerPrefab);
        cc.director.getCollisionManager().enabled = true;
    }

    setClouds(clouds: CloudData[], dt) {
        if (!Array.isArray(clouds)) {
            clouds = [];
        }
        if (clouds && clouds.length > 0) {
            for (let i = this.clouds.length; i < clouds.length; i++) {
                this.clouds.push(getFromCloudsPool(this.cloudPrefab));
            }
            let cloudNum = clouds.length - 5;
            clouds.forEach((cloud, i) => {
                const cloudView = this.clouds[i].getComponent(Cloud);
                if (i > cloudNum) {
                    const {x, y} = this.convertPosition(cloud.x, cloud.y);
                    cloudView.node.parent = this.node;
                    if (x + cloud.offset >= (this.maxX * this.tileSize + this.cloudSize / 2)) {
                        // 销毁节点
                        cloudView.node.parent = null;
                    } else {
                        cloud.offset += cloud.speed * dt;
                        cloudView.initCloud(x + cloud.offset, y);
                    }
                } else {
                    // 销毁节点
                    cloudView.node.parent = null;
                }
            });
        }
    }

    setPlayers(playerArr: PlayerData[]) {
        if (!Array.isArray(playerArr)) {
            playerArr = [];
        }
        this.players.splice(playerArr.length).forEach(player => removeToPlayerPool(player));
        for (let i = this.players.length; i < playerArr.length; i++) {
            this.players.push(getFromPlayersPool(this.playerPrefab));
        }
        playerArr.forEach((player, i) => {
            if (this.players[i]){
                const playerView = this.players[i].getComponent(Player);
                playerView.node.parent = this.node;
                if(player.hp == 0 && player.isShoot) {
                    let tempPlayer = frameSyncPlayerInitList.players.find(p => p.playerId == player.playerId);
                    player.hp = global.planeMaxHp;
                    player.x = tempPlayer.x;
                    player.y = tempPlayer.y;
                }
                playerView.initPlayer(player);
            }
        });
    }

    setBullet(bulletData: BulletData) {
        let bullet = this.node.getChildByName(bulletData.bulletId.toString());
        if(bullet) {
            console.log('---------移动子弹------');
            let bulletView = bullet.getComponent(Bullet);
            bulletView.updatePos(bulletData);
        } else {
            // 若为刚销毁的子弹，说明已经发生过碰撞，无需再创建了
            if(!destroyedBulletSet.has(bulletData.bulletId)) {
                console.log('---------创建子弹------');
                bullet = cc.instantiate(this.bulletPrefab);
                bullet.parent = this.node;
                let bulletView = bullet.getComponent(Bullet);
                bulletView.initBullet(bulletData);
            }
        }
    }

    destroyBullet(bulletId: string) {
        let bullet = this.node.getChildByName(bulletId);
        if(bullet) {
            let bulletView = bullet.getComponent(Bullet);
            bulletView.destroyBullet();
        }
    }

    convertPosition(mapX: number, mapY: number) {
        const x = mapX * this.tileSize + this.tileSize / 2;
        const y = mapY * this.tileSize + this.tileSize / 2;
        return {x, y};
    }

}
